internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,443 @@
+
+	// networkLibrary is the underlying network library. Can be used for testing.
+	networkLib model.UnderlyingNetworkLibrary
+}
+
+func (c *Config) repetitions() int64 {
+	if c.Repetitions > 0 {
+		return c.Repetitions
+	}
+	return 10
+}
+
+func (c *Config) port() string {
+	if c.Port != 0 {
+		return strconv.FormatInt(c.Port, 10)
+	}
+	return "443"
+}
+
+func (c *Config) networkLibrary() model.UnderlyingNetworkLibrary {
+	if c.networkLib != nil {
+		return c.networkLib
+	}
+	return &netxlite.TProxyStdlib{}
+}
+
+// TestKeys contains the experiment results.
+type TestKeys struct {
+	Domain      string        `json:"domain"`
+	Pings       []*SinglePing `json:"pings"`
+	Repetitions int64         `json:"repetitions"`
+}
+
+// SinglePing is a result of a single ping operation.
+type SinglePing struct {
+	ConnIdDst         string                         `json:"conn_id_dst"`
+	ConnIdSrc         string                         `json:"conn_id_src"`
+	Failure           *string                        `json:"failure"`
+	Request           *model.ArchivalMaybeBinaryData `json:"request"`
+	RequestTime       string                         `json:"request_time"`
+	Response          *model.ArchivalMaybeBinaryData `json:"response"`
+	ResponseTime      string                         `json:"response_time"`
+	SupportedVersions []uint32                       `json:"supported_versions"`
+}
+
+// Measurer performs the measurement.
+type Measurer struct {
+	config Config
+}
+
+// ExperimentName implements ExperimentMeasurer.ExperimentName.
+func (m *Measurer) ExperimentName() string {
+	return testName
+}
+
+// ExperimentVersion implements ExperimentMeasurer.ExperimentVersion.
+func (m *Measurer) ExperimentVersion() string {
+	return testVersion
+}
+
+// requestInfo contains the information of a sent ping request.
+type requestInfo struct {
+	t     time.Time
+	raw   []byte
+	dstID string
+	srcID string
+	err   error
+}
+
+// responseInfo contains the information of a received ping reponse.
+type responseInfo struct {
+	t        time.Time
+	raw      []byte
+	dstID    string
+	versions []uint32
+	err      error
+}
+
+// sender sends a ping requests to the target hosts every second
+func (m *Measurer) sender(
+	ctx context.Context,
+	pconn model.UDPLikeConn,
+	destAddr *net.UDPAddr,
+	out chan<- requestInfo,
+	sess model.ExperimentSession,
+) {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	i := 0
+	for i < int(m.config.repetitions()) {
+		select {
+		case <-ctx.Done():
+			return // user aborted or timeout expired
+
+		case sendTime := <-ticker.C:
+			packet, dstID, srcID := buildPacket()     // build QUIC Initial packet
+			_, err := pconn.WriteTo(packet, destAddr) // send Initial packet
+
+			sess.Logger().Infof("PING %s", destAddr)
+
+			// propagate send information, including errors
+			out <- requestInfo{raw: packet, t: sendTime, dstID: hex.EncodeToString(dstID), srcID: hex.EncodeToString(srcID), err: err}
+			i += 1
+
+		default:
```

Comment:
```suggestion
```

Having a `default` arm here strikes me as incorrect. It means you are not blocked in either of the two arms of the `select` and instead you are busy looping. I think you should double check the CPU utilization during the experiment and then remove this `default` arm of the `select` statement.